### PR TITLE
[ci] function to declare projects

### PR DESCRIPTION
**Pull request by gares**
This PR is not only about conciseness, but also to make the list of projects first class.
In this way we can fix the `notify-upstream-pins.sh` script after #13598 which removes the `.bat` file previously used.
I believe this script should be moved to the platform (see the FIXME) but for now we just repair it.

### CODE REVIEW

**Comment by ejgallego** [comment]
Thanks @gares , indeed that was the idea at the beginning but we hoped to write this infrastructure not in bash, but in OCaml or maybe Python. But sure bash works too, I'll wait for non-draft to provide more feedback but it should work.

A choice is actually to define a kind of json-style object that describes the project and just dump the metadata there, then we would have a bit more freedom to change the scripts implementation.

File: dev/ci/ci-common.sh
Comment (SkySkimmer) [inline dev/ci/ci-common.sh:58]:
```suggestion
    if ! is_in_projects "$1"; then
```

**Comment by gares** [comment]
it was a draft because it is broken, I must have done a silly mistake (even if locally it was kind of ok)

**Comment by gares** [comment]
> OCaml or maybe Python.

The current format is easy to parse, even via `grep ^project | cut -d ' '`, I think you can't do much better than this.

@SkySkimmer I see that you seem the only one to use the `overlay` function in you overlays. How is that?

**Review by SkySkimmer — COMMENTED** [review]
I guess people don't use the create_overlay script.

File: dev/ci/ci-common.sh
Comment (SkySkimmer) [inline dev/ci/ci-common.sh:61]: What's the point of this check?

Comment (gares) [reply dev/ci/ci-common.sh:61]: well, if you use the script I guess it's ok by construction.
if you write it by hand, you may make a typo, and the result is not as easy to debug as this error message.

**Comment by SkySkimmer** [comment]
There's an `overlay` user here https://github.com/coq/coq/pull/13537/files :fire:

**Comment by ejgallego** [comment]
> I guess people don't use the create_overlay script.

What do you mean? I do use it.

**Comment by gares** [comment]
now always:
```
gares@ollypat:~/COQ/master$ cat dev/ci/user-overlays/12611-ejgallego-record+refactor.sh
if [ "$CI_PULL_REQUEST" = "12611" ] || [ "$CI_BRANCH" = "record+refactor" ]; then

    elpi_CI_REF=record+refactor
    elpi_CI_GITURL=https://github.com/ejgallego/coq-elpi

#    mtac2_CI_REF=record+refactor
#    mtac2_CI_GITURL=https://github.com/ejgallego/Mtac2

fi
```
(I'm also one of the bad guys)

Maybe my question is: shouldn't the create_overlay script just write something like
```
overlay project url branch prnumber
```
in the file? (and have the if in common?)

**Comment by SkySkimmer** [comment]
>not always:

That overlay may be older than the `overlay` function

**Comment by ejgallego** [comment]
> That overlay may be older than the `overlay` function

Yup, that's the reason; I always use the script.

**Comment by gares** [comment]
> Maybe my question is: shouldn't the `create_overlay` script just write something like
> 
> ```
> overlay project url branch prnumber prbranch
> ```
> 
> in the file? (and have the `if` mess in `ci-common.sh`?)

And what about this?

**Comment by SkySkimmer** [comment]
>And what about this?

I don't care either way

**Comment by gares** [comment]
> I don't care either way

done, please review.

File: dev/ci/ci-basic-overlay.sh
Comment (gares) [inline dev/ci/ci-basic-overlay.sh:81]: It's a bit off topic for this PR, but still: @RalfJung @yforster is this repository still relevant? I've seen a repo for autosubst2 on github recently, maybe we should track that one?

Comment (Zimmi48) [reply dev/ci/ci-basic-overlay.sh:81]: Yes, the two repositories are still relevant.

Comment (RalfJung) [reply dev/ci/ci-basic-overlay.sh:81]: Yeah, I am going to delete my fork at some point. The master branch of https://github.com/coq-community/autosubst is the version that should be used.

File: dev/ci/ci-common.sh
Comment (SkySkimmer) [inline dev/ci/ci-common.sh:65]:
```suggestion
      if ! is_in_projects "$project"; then
```

**Review by SkySkimmer — APPROVED** [review]
Should I be concerned about the windows job failing?

File: dev/ci/ci-basic-overlay.sh
Comment (gares) [reply dev/ci/ci-basic-overlay.sh:81]: I'll update this thanks

**Comment by gares** [comment]
I don't think the windows failure is related. Anyway, #13598 should superseed it

**Comment by Zimmi48** [comment]
Windows is failing everywhere. See https://coq.zulipchat.com/#narrow/stream/237656-Coq-devs.20.26.20plugin.20devs/topic/Windows.20CI

**Comment by gares** [comment]
ping

**Comment by SkySkimmer** [comment]
@coqbot: merge now

File: dev/ci/user-overlays/README.md
Comment (Zimmi48) [inline dev/ci/user-overlays/README.md:21]: The text above about the file containing very simple logic should probably have been updated as well.

File: dev/ci/user-overlays/README.md
Comment (Zimmi48) [inline dev/ci/user-overlays/README.md:30]: The branching conventions section recommends using the same branch name for the overlay and the PR. It could have been updated to specify that in this case you avoid one optional argument to the `overlay` function.

File: dev/ci/user-overlays/README.md
Comment (Zimmi48) [inline dev/ci/user-overlays/README.md:28]: If what is referred from here is just this:

```
# overlay <project> <giturl> <ref> <prnumber> [<prbranch>]
#   creates an overlay for project using a given url and branch which is
#   active for prnumber or prbranch. prbranch defaults to ref.
```

Then I don't think it's very wise to refer to the code instead of updating the (almost complete) doc in this file.

**Review by Zimmi48 — COMMENTED** [review]
Sorry for having only looked at this after the merge.

BTW, expect some failures in the PRs that were already using the `overlay` function but without the last mandatory argument (the PR number).

**Comment by gares** [comment]
OK, I'll prepare another pr

**Review by ejgallego — COMMENTED** [review]
Nice work!

**Pull request merged**
